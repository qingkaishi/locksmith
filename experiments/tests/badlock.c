/* Should cause a non-linearity warning, since lock will be
   initialized twice. */
#include <pthread.h>

pthread_mutex_t lock;
int shared;

void *run(void *arg) {
  pthread_mutex_lock(&lock);
  shared++;
  pthread_mutex_unlock(&lock);
  return NULL;
}
void doit(int *j) {
  pthread_t t1;
  pthread_mutex_init(&lock, NULL);
  pthread_create(&t1, NULL, run, j);
}

int main() {
  int i = 0;
  doit(&i);
  doit(&i);
  return 1;
}
