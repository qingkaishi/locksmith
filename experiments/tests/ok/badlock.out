************************* STARTING *************************

shared: &shared:badlock.c:6

Warning: Possible data race: &shared:badlock.c:6 is protected by non-linear or concrete lock(s):
  lock:badlock.c:5
  non linear concrete lock::-1 -> doit badlock.c:22
  non linear concrete lock::-1 -> doit badlock.c:23
  non linear concrete lock:badlock.c:16
 references:
  dereference of &shared:badlock.c:6 at badlock.c:10
    &shared:badlock.c:6
  locks acquired:
    lock:badlock.c:5
    non linear concrete lock::-1 -> doit badlock.c:22
    non linear concrete lock::-1 -> doit badlock.c:23
    non linear concrete lock:badlock.c:16
  in: main at badlock.c:20  -> badlock.c:22
  in: main at badlock.c:20  -> badlock.c:23


*************************** DONE ***************************
